src/parse.rs

```diff
@@ -0,0 +1,2602 @@
+                self.next_tok();
+                self.parse_module_const(false)
+            }
+            (Some((_, Tok::Pub, _)), Some((_, Tok::Const, _))) => {
+                self.next_tok();
+                self.next_tok();
+                self.parse_module_const(true)
+            }
+
+            // External Type
+            (Some((start, Tok::External, _)), Some((_, Tok::Type, _))) => {
+                self.next_tok();
+                self.next_tok();
+                self.parse_external_type(start, false)
+            }
+            // External Function
+            (Some((start, Tok::External, _)), Some((_, Tok::Fn, _))) => {
+                self.next_tok();
+                self.next_tok();
+                self.parse_external_fn(start, false)
+            }
+            // Pub External type or fn
+            (Some((start, Tok::Pub, _)), Some((_, Tok::External, _))) => {
+                self.next_tok();
+                self.next_tok();
+                match self.next_tok() {
+                    Some((_, Tok::Type, _)) => self.parse_external_type(start, true),
+                    Some((_, Tok::Fn, _)) => self.parse_external_fn(start, true),
+                    _ => {
+                        self.next_tok_unexpected(vec!["A type of function definition".to_string()])?
+                    }
+                }
+            }
+
+            // function
+            (Some((start, Tok::Fn, _)), _) => {
+                self.next_tok();
+                self.parse_function(start, false, false)
+            }
+            (Some((start, Tok::Pub, _)), Some((_, Tok::Fn, _))) => {
+                self.next_tok();
+                self.next_tok();
+                self.parse_function(start, true, false)
+            }
+
+            // Custom Types, and Type Aliases
+            (Some((start, Tok::Type, _)), _) => {
+                self.next_tok();
+                self.parse_custom_type(start, false, false)
+            }
+            (Some((start, Tok::Pub, _)), Some((_, Tok::Opaque, _))) => {
+                self.next_tok();
+                self.next_tok();
+                self.expect_one(&Tok::Type)?;
+                self.parse_custom_type(start, true, true)
+            }
+            (Some((start, Tok::Pub, _)), Some((_, Tok::Type, _))) => {
+                self.next_tok();
+                self.next_tok();
+                self.parse_custom_type(start, true, false)
+            }
+
+            (t0, _) => {
+                self.tok0 = t0;
+                Ok(None)
+            }
+        };
+        tracing::trace!("Statement Parsed: {:?}", statement);
+        statement
+    }
+
+    //
+    // Parse Expressions
+    //
+
+    // expmples:
+    //   unit
+    //   unit op unit
+    //   unit op unit pipe unit(call)
+    //   unit op unit pipe unit(call) pipe unit(call)
+    fn parse_expression(&mut self) -> Result<Option<UntypedExpr>, ParseError> {
+        // uses the simple operator parser algorithm
```

Comment:
Realizing I called it Simple Operator Parser in the code but it's actually called "Simple Precedence Parser".
Here's a somewhat dense explanation on wikipedia https://en.wikipedia.org/wiki/Simple_precedence_parser

Essentially, it's a shift-reduce, bottom-up parser with the only goal of correctly applying precedence.
I use 2 stacks instead of the 1 because I think it's easier to reason about and type.
A stack for storing operators that have been seen but not yet reduced `opstack`
A stack for storing operands (expressions) `estack` which may be bare values like "1" or "0.5" or sub expressions like "1 * 5"

While reading the input stream: 
if you see an operand put it at the top of the estack
if you see and operator compare it to to the top of the opstack and do some stuff
when you get to the end of the input try to reduce the rest of the operators and expressions

Happy to go over the algorithm in more detail if you want.
Lemme know what kind of documentation / examples we want to put into the source!

Reply:
This is actually a great video explanation: https://www.youtube.com/watch?v=n5UWAaw_byw&t=648s

Reply:
That's great, thank you. Some of this info in the comment would be fab!